include/fast_ber/ber_types/Real.hpp

```diff
@@ -47,7 +125,610 @@ class Real
+        exponent = 0;
+    }
+
+    /* Finally add the mantissa sign and the exponent to the double bytes */
+    double_bytes[0] = (mantissa_sign << 7) | (exponent >> 4);
+    double_bytes[1] = (exponent << 4) | (double_bytes[1] & 0x0F);
+
+#ifdef ABSL_IS_LITTLE_ENDIAN
+    /* Reverse bytes if needed */
+    std::reverse(double_bytes.begin(), double_bytes.end() - 1);
+#endif
+
+    memcpy(&output, double_bytes.data(), sizeof(double));
+    return true;
+}
+
+/**
+ * Decodes a real decimal value
+ * @param input BER encoded real decimal value
+ * @param output decoded decimal value
+ * @return true on success, false on failure
+ */
+inline bool decode_real_decimal(absl::Span<const uint8_t> input, double& output) noexcept
+{
+    std::string iso_6093_value(reinterpret_cast<const char*>(&input[1]), input.length() - 1);
+
+    /* Check that string matches the iso 6093 NR1/NR2/NR3 forms */
+    if (std::regex_match(iso_6093_value, iso_6093_nr1) || std::regex_match(iso_6093_value, iso_6093_nr2) ||
+        std::regex_match(iso_6093_value, iso_6093_nr3))
+    {
+        /* Replace potential comma by a dot */
+        std::replace(iso_6093_value.begin(), iso_6093_value.end(), ',', '.');
+
+        return absl::SimpleAtod(iso_6093_value, &output);
+    }
+    else
+    {
+        return false;
+    }
+}
+
+inline size_t encode_real(absl::Span<uint8_t> output, double input) noexcept
+{
+    std::array<uint8_t, sizeof(double)> buffer;
+    std::memcpy(buffer.data(), &input, sizeof(double));
+
+#ifdef ABSL_IS_LITTLE_ENDIAN
+    /* Reverse bytes if needed */
+    std::reverse(buffer.begin(), buffer.end());
+#endif
+
+    unsigned int sign           = buffer[0] >> 7;
+    size_t       encoded_length = 0;
+    int          exponent       = (buffer[0] & 0x7F) << 4 | (buffer[1] >> 4);
+
+    /* Remove exponent */
+    buffer[1] &= 0x0F;
+
+    if (exponent == 0x7FF)
+    {
+        if (std::all_of(buffer.begin() + 1, buffer.end(), [](uint8_t elem) { return elem == 0; }))
+        {
+            if (sign == 0)
+            {
+                /* Plus infinity */
+                output[0] = 0x40;
+            }
+            else
+            {
+                /* Minus infinity */
+                output[0] = 0x41;
+            }
+        }
+        else
+        {
+            /* Not a number value */
+            output[0] = 0x42;
+        }
+
+        return 1;
+    }
+    else if (exponent == 0)
+    {
+        if (std::all_of(buffer.begin() + 1, buffer.end(), [](uint8_t elem) { return elem == 0; }))
+        {
+            if (sign == 0)
+            {
+                /* Zero value - nothing to do */
+                return 0;
+            }
+            else
+            {
+                /* Minus zero */
+                output[0] = 0x43;
+                return 1;
+            }
+        }
+        else
+        {
+            /* Denormalized value */
+            buffer[8] = 0;
+            for (size_t i = 1; i < buffer.size(); i++)
+            {
```

Comment:
checking for unistd.h... In file included from /home/styler/git/kurkpit/fast_ber/test/ber_types/RealTest.cpp:1:
/home/styler/git/kurkpit/fast_ber/src/../include/fast_ber/ber_types/Real.hpp: In function ‘size_t fast_ber::encode_real(absl::Span<unsigned char>, double)’:
/home/styler/git/kurkpit/fast_ber/src/../include/fast_ber/ber_types/Real.hpp:576:75: warning: iteration 6 invokes undefined behavior [-Waggressive-loop-optimizations]
                 buffer[i]                         = (current_byte_shifted | buffer[i + 1]) >> 7;
                                                     ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/home/styler/git/kurkpit/fast_ber/src/../include/fast_ber/ber_types/Real.hpp:573:34: note: within this loop
             for (size_t i = 1; i < buffer.size(); i++)
                                ~~^~~~~~~~~~~~~~~

Reply:
Fixed